Feature/describe prior to write

detule opened the pull request:
Hi @jimhester @krlmlr 

Please consider this patch-set as potentially a more robust alternative to https://github.com/r-dbi/odbc/issues/70.

* With the current solution (master) FreeTDS users face truncation when inserting values longer than 255 characters.
* An additional benefit could be performance enhancement when writing to wide tables, as this approach makes less round-trip calls to the server.

Looking forward to hearing your feedback.

jimhester commented:
Thanks for working on this!

It looks like this change is causing the tests to fail when using PostgreSQL, would you be able to look into this?

detule commented:
Will do - thanks.

detule commented:
@jimhester Should be ready for a second look.

The R-devel pipeilne failure seems to be [unrelated to this patch-set](https://travis-ci.community/t/new-warning-in-r-devel-about-the-checkbashisms-script/5551).

jimhester commented on R/Table.R:
```diff
@@ -62,6 +71,9 @@ odbc_write_table <-
       rs <- OdbcResult(conn, sql)
 
+      if (!is.null(fieldDetails) && nrow(fieldDetails))
+        result_describe_parameters(rs@ptr, fieldDetails)
```
Can you put braces around this conditional?

detule replied:
Will do - thanks.

jimhester commented on R/Table.R:
```diff
@@ -50,6 +50,15 @@ odbc_write_table <-
+      tryCatch({
+        details <- odbcConnectionColumns(conn, name)
+        datails <- details[match(names(values), details$column_name)]
+        details[, c("ordinal_position", "data_type", "column_size", "decimal_digits")]
```
This call has no effect, did you mean to assign it back to details?

detule replied:
Jim - the output of the `tryCatch` block is assigned to `fieldDetails` -  the line you quoted there is what gets assigned in the success code-path.

Let me know if you think there is a better approach.

jimhester replied:
It was due to the indentation style used, I would normally put the `tryCatch()` line on the same line as the assignment.

detule replied:
Will update - thanks.

jimhester commented on R/Table.R:
```diff
@@ -50,6 +50,15 @@ odbc_write_table <-
+    fieldDetails <-
+      tryCatch({
+        details <- odbcConnectionColumns(conn, name)
+        datails <- details[match(names(values), details$column_name)]
```
This could just be

```r
datails <- details[names(values)]
```

But do we need to reorder the columns at all?

detule replied:
Jim, the intention here was to:
* Subset the rows of the output of SQLColumns to those (columns) we are writing to / we need to describe.
* I thought some care was needed with the ordering - AFAICT there is no guarantee that the columns in `values` (and therefore the parameters in the INSERT statement) are ordered the same way as the columns in the table we are writing to (i.e. the rows in `details`).

Having said that, I think part of the confusion is the use of `ordinal_position` to index the parameters we are describing - this seems incorrect.  After re-ordering the rows in `details` to match the column order in `values` as is done here, the index that gets passed `nanodbc::statement::describe_parameters` should be `1:ncol(values)`

jimhester commented on R/Connection.R:
```diff
@@ -69,6 +69,101 @@ setClass(
+      "sql_datetime_subtype", "char_octet_length", "ordinal_position",
+      "nullable")
+    detail <-
+      detail[, c(5, 4, 3, 1, 6, 2, 7, 8, 9, 10, 17, 11, 12, 13, 14, 15, 16)]
```
Indexing by number seems too brittle, could we reorder by name first, then rename the columns if necessary.

detule replied:
Will change - thanks.

jimhester commented on R/Connection.R:
```diff
@@ -69,6 +69,101 @@ setClass(
+      detail[, c(5, 4, 3, 1, 6, 2, 7, 8, 9, 10, 17, 11, 12, 13, 14, 15, 16)]
+
+    if (!is.null(column_name))
+      detail <- subset(detail, detail$column_name == column_name)
```
Could we add braces here,

Also generally it is best to avoid using subset inside functions, the type of non-standard evaluation in `subset()` is not robust enough.

`detail <- detail[detail$column_name == column_name, ]` is the equivalent.

Also it is somewhat surprising that you would need to do an additional filtering, if you search for a specific column in `SQLColumns` only that column should be returned.

detule replied:
Will change - thanks.

detule commented:
@jimhester Thanks for the feedback!

* Patched conditional braces, numeric column indexing
* Fixed bug with pairing table column descriptions with INSERT statement parameters.  Added a test.

`test_roundtrip` is perhaps not the best home for this test - happy to move it to a more appropriate location if you have any suggestions.

jimhester commented on R/Connection.R:
```diff
@@ -69,6 +69,98 @@ setClass(
+      "decimal_digits", "numeric_precision_radix", "nullable", "remarks",
+      "column_default", "sql_data_type", "sql_datetime_subtype",
+      "char_octet_length", "ordinal_position")]
+    names(detail)[c(1, 2, 4, 6, 13)] <- c("table_cat", "table_schem",
```
I think I would prefer to _not_ rename. And do we need to reorder the columns?

detule replied:
Jim my original thinking here was to make the output as close to the output of the ODBC `SQLColumns` function as possible.   It seemed to me that, since we are offering the end-user the ability to write their own implementation, then it is more natural to anchor on the field naming of the ODBC's `SQLColumns` API, than on the choices made when implementing the internal `connection_sql_columns`.

If you also think that's worthwhile then we either re-name the columns here, or change the output of `connection_sql_columns`.  The latter seemed less desirable since even though internal, I wasn not sure if there aren't folks that have developed dependencies.

The re-ordering was more thinking in the same direction - making the output align with what I can see when I execute `sp_columns` in my SQL Server client for example.

I don't feel strongly here - happy to make this just a wrapper around `connection_sql_columns` if you think that's the best way forward.

jimhester replied:
I would prefer it to just be a wrapper around `connection_sql_columns`

jimhester commented:
I would prefer a separate test outside of the roundtrip that creates a table and appends to it.

detule commented:
@jimhester Thanks again - appreciate the comments.

Let me know if I missed anything.

jimhester commented:
![Thanks!! You are da bomb!](https://media.giphy.com/media/eBqUy3ik6P2vK/giphy.gif)

detule commented:
Jim: thank you very much - I know it was a chunky PR, and I appreciate the thoughtful comments.

Any thoughts on when this might make its way to CRAN?

jimhester commented:
I was aiming for a CRAN release on odbc next week sometime.
